core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java

```diff
@@ -620,6 +699,17 @@ private void deleteAndVerifyGroupConfigValue(Admin client,
         verifyGroupConfig(client, groupName, defaultConfigs);
     }
 
+    private void deleteAndVerifyClientMetricsConfigValue(Admin client,
+                                                         String clientMetricsName,
+                                                         Map<String, String> defaultConfigs) throws Exception {
```

Comment:
Yes, I will do this. I created https://issues.apache.org/jira/browse/KAFKA-17516 to reflect the problem with the handling of default configs for client metrics, but that's not yet resolved. I'll do that separately.

Reply:
I don't think that it is necessary to check for the entity name on line 978 for client-metrics. The effect is that the tool describes *all* entities of that type. It cannot do that for broker-logger. It also cannot do that for groups, so I've made the change for groups. Really, there ought to be a tiny KIP like KIP-1000 to fix this for groups.

Reply:
>  The effect is that the tool describes all entities of that type.

Pardon me, what happens if user pass `--describe --entity-type client-metrics --entity-default` ? I assumes server will reject such request.
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ConfigHelper.scala#L141

If my assumption is correct, we should do the check like `broker-logger` and `groups`.

Reply:
> Really, there ought to be a tiny KIP like KIP-1000 to fix this for groups.

Did you mean to return "default configs"  for groups?

Reply:
Ah, there are multiple points here. Thanks for your patience.

If you do `--describe --entity-type topics`, it lists the configs for all topics. If you do `--describe --entity-type client-metrics`, it lists configs for all client metrics resources. If you do `--describe --entity-type groups`, it would ideally list all of the group config resources, but there is no RPC capable of doing that so instead it lists the consumer groups with configs.

The `--entity-default` situation is tricky. None of topics, groups or client-metrics works with `--entity-default`. Making it required to specify an entity name for these types would break the ability to list the configs for all resources. `--entity-default` should only be permitted for clients, users, ips and brokers, but there's no explicit check for this.

I will revert my change for group with no entity name, and add a check for `--entity-default` to make sure it's only specified with entity types that genuinely support it.

There is also the point about default values for the configs themselves. So, if I have a group G1 with just `consumer.heartbeat.interval.ms` defined, then this is the output:

```
$ bin/kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type groups --entity-name G1
Dynamic configs for group G1 are:
  consumer.heartbeat.interval.ms=5000 sensitive=false synonyms={DYNAMIC_GROUP_CONFIG:consumer.heartbeat.interval.ms=5000}

$ bin/kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type groups --entity-name G1 --all
All configs for group G1 are:
  consumer.heartbeat.interval.ms=5000 sensitive=false synonyms={DYNAMIC_GROUP_CONFIG:consumer.heartbeat.interval.ms=5000}
  consumer.session.timeout.ms=45000 sensitive=false synonyms={}
```

Currently, client-metrics does not behave the same (and it should). If a client-metrics resource defines a subset of the allowed configs, the `--all` does not display the default values for the missing configs. I'll address that in KAFKA-17516.

Reply:
@AndrewJSchofield thanks for all your sharing!

Reply:
> consumer.session.timeout.ms=45000 sensitive=false synonyms={}

Should the value of `synonyms` be `synonyms={DEFAULT_CONFIG:group.consumer.session.timeout.ms=45000}`?

Reply:
That would imply that the broker config `group.consumer.session.timeout.ms` provided the default. KIP-848 does describe such a config, but does not explicitly specify that the default should be taken from the broker config. I think that probably it should and this is a symptom of that problem.

As an aside, we are doing exactly this in https://github.com/apache/kafka/pull/17070 which is creating a group config with a broker default.